src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -1069,6 +1080,23 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     {
                         GetEmitter()->emitIns_R_R(ins, emitSize, targetReg, op1Reg, opt);
                     }
+                    else if (HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrin.id) && intrin.op2->isContained())
```

Comment:
I didn't understand which scenario this is trying to handle?

Reply:
This is for `AddSequentialAcross` (`fadda`). This is the only intrinsic amongst the XAcross we are optimising where the optimisation is possible when the csel is at op2 and it meets other conditions ( zero as the third op of csel).
In the above list in the description, we can optimise item 2. (csel as second op of across) but not 3. (csel as the first operand of across).
For item 3, we get the wrong result if we do the current optimisation.

Reply:
can you show the codegen for 2 and 3?

Reply:
For case 2 (with optimisation)

<details>
<summary> Code example </summary>

```C#
public static Vector<double> Sve_AddSequentialAcross1()
{
    var vec1 = Vector.Create((double)42);
    var vec2 = Vector.Create((double)43);
    var mask = Vector.Create((double)1);
    return Sve.AddSequentialAcross(vec2, Sve.ConditionalSelect(mask, vec1, Vector<double>.Zero));
}

```
</details>


```
...
G_M19924_IG02:  ;; offset=0x0018
            ldr     d0, [@RWD00]
            stp     d0, d0, [fp, #0x30]	// [V04 tmp2], [V04 tmp2+0x08]
            ldr     q0, [fp, #0x30]	// [V04 tmp2]
            ldr     d16, [@RWD08]
            stp     d16, d16, [fp, #0x20]	// [V09 tmp7], [V09 tmp7+0x08]
            ldr     q16, [fp, #0x20]	// [V09 tmp7]
            fmov    d17, #1.0000
            stp     d17, d17, [fp, #0x10]	// [V14 tmp12], [V14 tmp12+0x08]
            ldr     q17, [fp, #0x10]	// [V14 tmp12]
            ptrue   p0.d
            cmpne   p0.d, p0/z, z17.d, #0
            fadda   v16.d, p0, v16.d, z0.d
            mov     v0.16b, v16.16b

G_M19924_IG03:  ;; offset=0x004C
            ldp     fp, lr, [sp], #0x40
            ret     lr
```


For case 3 (no optimisation)

<details>
<summary> Code example </summary>

```C#
public static Vector<double> Sve_AddSequentialAcross1()
{
    var vec1 = Vector.Create((double)42);
    var vec2 = Vector.Create((double)43);
    var mask = Vector.Create((double)1);
    return Sve.AddSequentialAcross(Sve.ConditionalSelect(mask, vec1, Vector<double>.Zero), vec2);
}

```
</details>

```
...
G_M38135_IG02:  ;; offset=0x0018
            ldr     d0, [@RWD00]
            stp     d0, d0, [fp, #0x30]	// [V04 tmp2], [V04 tmp2+0x08]
            ldr     q0, [fp, #0x30]	// [V04 tmp2]
            ldr     d16, [@RWD08]
            stp     d16, d16, [fp, #0x20]	// [V09 tmp7], [V09 tmp7+0x08]
            ldr     q16, [fp, #0x20]	// [V09 tmp7]
            fmov    d17, #1.0000
            stp     d17, d17, [fp, #0x10]	// [V14 tmp12], [V14 tmp12+0x08]
            ldr     q17, [fp, #0x10]	// [V14 tmp12]
            ptrue   p0.d
            cmpne   p0.d, p0/z, z17.d, #0
            movi    v17.4s, #0
            sel     z0.d, p0, z0.d, z17.d
            ptrue   p0.d
            fadda   v0.d, p0, v0.d, z16.d

G_M38135_IG03:  ;; offset=0x0054
            ldp     fp, lr, [sp], #0x40
            ret     lr
```

Reply:
For case 3, there are multiple things I am spotting odd:

1. `movi v17.4s, #0` : We want to move 0 to scalable register I assume because that's what we are using in next `sel` instruction, so shouldn't this be `fmov z17.s, #0`?
2. If I see `AddSequentialAcross(initial, value)` and my understanding is `initial` is the source / destination SIMD register, but above, when we do `fadda`, we are giving it `v0.d` as source register which is populated using `sel` in scalable register. When we support VL agnostic, won't we lose data? I am thinking this falls under the same category as [LASTB](https://docsmirror.github.io/A64/2023-06/lastb_v_p_z.html) where destination is SIMD & FP register instead of scalable register. (reference: https://github.com/dotnet/runtime/pull/103847#discussion_r1658060716). If yes, then we should rethink about this API or perhaps, if someone uses `ConditionalSelect()`, here we will just clobber the data and operate on SIMD registers.

Reply:
In case 1., it seems to be optimised `fmov` to `movi` as we are moving zero.
Case 2., you're right. We may need to update the API calls to reflect that the initial and return value is of type `Vector128` while the op2 is of type `Vector<T>`.

Reply:
@tannergooding - what is your thoughts on this?